### DISCUSSION TITLE
How to generate all paths between two nodes (including repeated nodes)

### DISCUSSION THREAD

**Discussion by jeertmans** [networkx/networkx#7094] [Q&A] [answered]
Hello! I could not find information about this on the Internet, but I was wondering if there was an already existing implementation of `all_simple_paths`, but including repeated nodes?

**Comment by MridulS** [comment]
> existing implementation of all_simple_paths, but including repeated nodes?

We currently only have a function to calculate [number_of_walks](https://networkx.org/documentation/latest/reference/algorithms/generated/networkx.algorithms.walks.number_of_walks.html). (It's a nice property of the adjacency matrix :) ) but it's doesn't calculate the explicit paths.

The right keyword to look for here is a `walk`, a `path` can not have repeated nodes.

**Reply by jeertmans** [reply]
Oh thanks! I thought it was the « Simple’s keyword that was denoting the non duplicate nodes

**Reply by MridulS** [reply]
ahhhhh, yes you are right, my bad. The "simple" there does denote the non duplicate node (I was thinking about non duplicate edges). But we don't have anything specific for `all_paths`. 

Do you have a MultiGraph (graph with multiple edges between 2 nodes)?

**Reply by jeertmans** [reply]
Actually I try to generate all possibles paths from one node to another, and I use a ones matrix as adjacency matrix (from which I built my graph). A path is valid a long as it does not contain loops (so a path does not contains two consecutive equal node ids).

**Reply by dschult** [reply]
There are a few extreme definitions of path, but the mainstream definitions, e.g. [wikipedia path](https://en.wikipedia.org/wiki/Path_(graph_theory)) says a path does not include duplicate nodes.  But it adds a qualifier "by most definitions".  But when considering a path as a sequence of edges, the edges can contain self-loops, so in this case the path can contain the same node more than once. But a "simple path" does not.

It sounds like your use-case has every node connected to every node (adjacency matrix is ones). In that case, aren't you looking for `itertools.product`?   Or look at other itertools options like `permutations` or `combinations`.

**Reply by jeertmans** [reply]
Yes it is like itertools.product, expect I need to remove all paths that contain two or more consecutive duplicate nodes. And if I want to disconnect, say, just two nodes in my graph, then it becomes even harder to implement this with itertools.product

**Reply by dschult** [reply, marked as answer]
Hopefully something like this could work. :)

```python
nodes = list(range(5))
source, target = 0, 1
not_connected = nx.empty_graph(nodes)
not_connected.add_edges_from([(1, 2), (0, 2)])

all_nonrepeated_node_paths = []
if not not_connected.has_edge(source, target):
    all_nonrepeated_node_paths.append((source, target))

other_nodes = [n for n in nodes if n not in (source, target)]
for N in range(len(other_nodes), 0, -1):
  maybe_paths = itertools.permutations(unused_nodes, N)
  for p in maybe_paths:
    # check first edge
    if not_connected.has_edge(source, p[0]):
        continue
    # check last edge
    if not_connected.has_edge(p[-1], target):
        continue
    # check other edges
    for u, v in itertools.pairwise(p):
        if not_connected.has_edge(u, v):
            break
    else:
        full_path = (source,) + p + (target,)
        all_nonrepeated_node_paths.append(full_path)

print(all_nonrepeated_node_paths)
```

**Reply by jeertmans** [reply]
Thanks for helping @dschult, I will try this out when I have some time :-)
